src/transformers/trainer.py

```diff
@@ -863,9 +753,36 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
                 state_dict = torch.load(os.path.join(self.state.best_model_checkpoint, WEIGHTS_NAME))
                 self.model.load_state_dict(state_dict)
 
+        self.control = self.callback_handler.on_train_end(self.args, self.state, self.control)
+
         return TrainOutput(self.state.global_step, tr_loss.item() / self.state.global_step)
 
-    def _save_training(self, model, trial, metrics=None):
+    def _maybe_log_save_evalute(self, tr_loss, model, trial, epoch):
```

Comment:
(nit) Think it's better to have the `if self.control.should_log:` statement before the two calls of `_maybe_log_save_evaluate()` and rename the function to `_log_save_evaluate()` => "maybe" in a function name sounds a bit scary to the reader IMO.